### DISCUSSION TITLE
How to configure token for private docker registry

### DISCUSSION THREAD

**Discussion by tschut** [renovatebot/renovate#20552] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.148.0

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

n/a

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate uses the token I configure without trying to go trough the oauth flow.

### What you tried so far.

We're running a private docker registry in an openshift environment. The following works:
```
$ oc login https://my-cluster --username=$username --password=$password
$ curl -k -H "authorization: Bearer $(oc whoami -t)" https://my-registry/v2/my-image/tags/list?n=10000
{"name":"my-image","tags":["5","latest","1","2","3","4.9","4.9.1"]}
```

I'm running renovate in a jenkins pipeline, and I'm passing the token to renovate like this:
```
sh script: 'oc login https://my-registry --username="$DOCKER_USERNAME" --password="$DOCKER_PASSWORD"'
env.OC_TOKEN = sh script: 'oc whoami -t', returnStdout: true
env.RENOVATE_HOST_RULES = '[{"hostType": "docker", "token": "$OC_TOKEN"}]'
```

I've verified that the token is set correctly. Renovate is also picking the hostrule up correctly. However, when running renovate it tries to get a token from `https://my-registry/openshift/token?service=&scope=repository:my-image:pull`. However, this fails with a 401. Also, it's not necessary as demonstrated by the `curl` request above, renovate could just use the supplied token.

What I'm not sure about is if this is a configuration error on my end, a bug in renovate, or weird behavior of openshift. Any help would be appreciated.

This is the function that performs the call to `/openshift/token?...`: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/docker/index.ts#L60. I'm wondering if it should get here it all if I've already supplied a bearer token in the hostrule.

### Relevant debug logs

<details><summary>Logs</summary>

```
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":30,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","baseBranch":"main","stats":{"managers":{"dockerfile":{"fileCount":1,"depCount":1}},"total":{"fileCount":1,"depCount":1}},"msg":"Dependency extraction complete","time":"2023-02-21T15:27:42.750Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","msg":"GET https://my-registry/openshift/token?service=&scope=repository:my-image:pull = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=12)","time":"2023-02-21T15:27:42.880Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","registryHost":"https://my-registry","dockerRepository":"my-image","msg":"Unauthorized docker lookup","time":"2023-02-21T15:27:42.880Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","err":{"name":"HTTPError","code":"ERR_NON_2XX_3XX_RESPONSE","timings":{"start":1676993262866,"socket":1676993262866,"lookup":1676993262869,"connect":1676993262870,"secureConnect":1676993262872,"upload":1676993262872,"response":1676993262877,"end":1676993262878,"phases":{"wait":0,"dns":3,"tcp":1,"tls":2,"request":0,"firstByte":5,"download":1,"total":12}},"message":"Response code 401 (Unauthorized)","stack":"HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/lib/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)","options":{"headers":{"user-agent":"RenovateBot/34.148.0 (https://github.com/renovatebot/renovate)","accept":"application/json","authorization":"***********","accept-encoding":"gzip, deflate, br"},"url":"https://my-registry/openshift/token?service=&scope=repository:my-image:pull","hostType":"docker","username":"","password":"","method":"GET","http2":false},"response":{"statusCode":401,"statusMessage":"Unauthorized","body":"","headers":{"content-type":"application/json","docker-distribution-api-version":"registry/2.0","content-length":"0","date":"Tue, 21 Feb 2023 15:27:42 GMT","set-cookie":["fbc1c0d076a66086a2bb1bf0f2ff0e36=df77a5838e7d54d0be56ad083b34d971; path=/; HttpOnly; Secure; SameSite=None"],"connection":"close"},"httpVersion":"1.1","retryCount":0}},"msg":"Response code 401 (Unauthorized)","time":"2023-02-21T15:27:42.881Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","msg":"Failed to get authHeaders for getTags lookup","time":"2023-02-21T15:27:42.881Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","dependency":"my-registry/my-image","packageFile":"Dockerfile","msg":"Failed to look up docker dependency my-registry/my-image","time":"2023-02-21T15:27:42.895Z","v":0}
{"name":"renovate","hostname":"oc4-jenkins-agent-renovate-rw0d0","pid":344,"level":20,"logContext":"C-DUX-qvAxU5tequM6Pyh","repository":"test/renovate-test","baseBranch":"main","msg":"PackageFiles.add() - Package file saved for base branch","time":"2023-02-21T15:27:42.895Z","v":0}
```

</details>

**Comment by secustor** [comment]
This will not work as single quotes prevent variable substitution. You will have to use `config.js` with `process.env.OC_TOKEN` instead.

**Reply by tschut** [reply]
I thought that the variable would still be substituted on the shell level as it's an environment variable, but I may be wrong. Either way, I tried your suggestion like so:
```groovy
script {
    sh script: 'oc login https://api.my-openshift-cluster --username="$DOCKER_USERNAME" --password="$DOCKER_PASSWORD"'
    env.OC_TOKEN = sh script: 'oc whoami -t', returnStdout: true
    sh script: 'set +x && echo "module.exports = { secrets: { OC_TOKEN: \'"$(oc whoami -t)"\' } };" > config.js && set -x'
}
```

And then extended my config with:
```
"hostRules": [
  {
    "matchHost": "https://my-registry",
    "token": "{{ secrets.OC_TOKEN }}"
  }
]
```

But I'm running into the same error.

It seems the auth flow on openshift is different? When I `curl` the /token endpoint with a Bearer token I also get a 401, even though the token is valid (which I checked by getting the image tags directly using said token). This issue would be fixed for me if I could somehow bypass the call to /token and have Renovate just use the supplied bearer token in the the call to the endpoint where it gets the tags.

Note that when I don't supply a bearer token, the call to /token returns `{"access_token":"anonymous","token":"anonymous"}`, and in subsequent calls Renovate will try to use "anonymous" as a token, which then obviously fails with a 401.

**Reply by viceice** [reply]
how would you login with the docker cli?

**Comment by tschut** [comment, marked as answer]
I stumbled upon the solution: it turns out that by supplying username/password combination instead of the token, where the password is the token I was using previously, everything works.

So, in my Jenkins file I now have:
```groovy
sh script: 'set +x && echo "module.exports = { secrets: { OC_USERNAME: \'"$(oc whoami)"\', OC_TOKEN: \'"$(oc whoami -t)"\' } };" > config.js && set -x'
```

And then in config.js:
```js
  "hostRules": [
    {
      "matchHost": "https://my-registry",
      "username": "{{ secrets.OC_USERNAME }}",
      "password": "{{ secrets.OC_TOKEN }}"
    }
  ]
```

I found out that when calling the /token endpoint with basic auth username:token, openshift will answer in the form: `{"access_token":"sha256~mytokenhere","token":"sha256~mytokenhere"}. In other words, it will "mirror" the token you provided in the basic auth header. Renovate will then use this token and everything works :)

This feels like an unnecessary step, but for now I'm just happy it works!

**Reply by viceice** [reply]
that's the same way the docker cli works.
